File: src/algorithms/sorting/distribution/bead-sort/bead-sort.test.ts
import { describe, it, expect } from "vitest";
import { beadSort } from "./sources/bead-sort.ts?fn";

describe("beadSort", () => {
  it("sorts an unsorted array", () => {
    expect(beadSort([64, 34, 25, 12, 22, 11, 90])).toEqual([11, 12, 22, 25, 34, 64, 90]);
  });

  it("handles an already sorted array", () => {
    expect(beadSort([1, 2, 3, 4, 5])).toEqual([1, 2, 3, 4, 5]);
  });

  it("handles a reverse-sorted array", () => {
    expect(beadSort([5, 4, 3, 2, 1])).toEqual([1, 2, 3, 4, 5]);
  });

  it("handles an array with duplicate values", () => {
    expect(beadSort([3, 1, 4, 1, 5, 9, 2, 6, 5])).toEqual([1, 1, 2, 3, 4, 5, 5, 6, 9]);
  });

  it("handles a single element array", () => {
    expect(beadSort([42])).toEqual([42]);
  });

  it("handles an empty array", () => {
    expect(beadSort([])).toEqual([]);
  });

  it("handles all identical elements", () => {
    expect(beadSort([7, 7, 7, 7])).toEqual([7, 7, 7, 7]);
  });

  it("handles an array with all zeros", () => {
    expect(beadSort([0, 0, 0])).toEqual([0, 0, 0]);
  });

  it("handles negative numbers by offsetting", () => {
    expect(beadSort([3, -1, 0, -5, 2])).toEqual([-5, -1, 0, 2, 3]);
  });

  it("does not mutate the original array", () => {
    const original = [3, 1, 2];
    const sorted = beadSort(original);
    expect(sorted).toEqual([1, 2, 3]);
    expect(original).toEqual([3, 1, 2]);
  });
});

File: src/algorithms/sorting/distribution/bead-sort/__tests__/BeadSort_test.java
public class BeadSort_test {
    public static void main(String[] args) {
        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{64, 34, 25, 12, 22, 11, 90}),
            new int[]{11, 12, 22, 25, 34, 64, 90}
        ) : "Test failed: sorts an unsorted array";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{1, 2, 3, 4, 5}),
            new int[]{1, 2, 3, 4, 5}
        ) : "Test failed: handles an already sorted array";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{5, 4, 3, 2, 1}),
            new int[]{1, 2, 3, 4, 5}
        ) : "Test failed: handles a reverse-sorted array";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{3, 1, 4, 1, 5, 9, 2, 6, 5}),
            new int[]{1, 1, 2, 3, 4, 5, 5, 6, 9}
        ) : "Test failed: handles an array with duplicate values";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{42}),
            new int[]{42}
        ) : "Test failed: handles a single element array";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{}),
            new int[]{}
        ) : "Test failed: handles an empty array";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{7, 7, 7, 7}),
            new int[]{7, 7, 7, 7}
        ) : "Test failed: handles all identical elements";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{0, 0, 0}),
            new int[]{0, 0, 0}
        ) : "Test failed: handles array with all zeros";

        assert java.util.Arrays.equals(
            BeadSort.beadSort(new int[]{3, -1, 0, -5, 2}),
            new int[]{-5, -1, 0, 2, 3}
        ) : "Test failed: handles negative numbers";

        int[] original = new int[]{3, 1, 2};
        int[] sorted = BeadSort.beadSort(original);
        assert java.util.Arrays.equals(sorted, new int[]{1, 2, 3}) : "Test failed: sorted result";
        assert java.util.Arrays.equals(original, new int[]{3, 1, 2}) : "Test failed: original not mutated";

        System.out.println("All tests passed!");
    }
}
